internal/modproxy/proxy.go

```diff
@@ -0,0 +1,177 @@
+package modproxy
+
+import (
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path"
+	"strings"
+
+	"golang.org/x/mod/modfile"
+	"golang.org/x/mod/semver"
+)
+
+// Proxy wraps a Getter and a list of proxy URLs to provide the required module proxy operations
+type Proxy struct {
+	g       Getter
+	proxies []string
+}
+
+// New returns a Proxy instance that will use g to execute HTTP requests against the module proxies
+// in urls.
+func New(g Getter, urls ...string) Proxy {
+	if len(urls) == 0 {
+		urls = getModProxies()
+	}
+	return Proxy{
+		g:       g,
+		proxies: urls,
+	}
+}
+
+// NewFromEnv returns a Proxy instance that will use g to execute HTTP requests against the module proxies
+// configured in the system environment
+func NewFromEnv(g Getter) Proxy {
+	return New(g)
+}
+
+// GetCurrentVersion returns the highest known version of the specified module, as returned by list of
+// module proxies configured on p.
+func (p Proxy) GetCurrentVersion(mod string, includePrerelease bool) (string, error) {
+	versions, err := p.GetModuleVersions(mod)
+	if err != nil {
+		return "", err
+	}
+	latest := versions[0]
+	for _, v := range versions {
+		if semver.Compare(v, latest) > 0 && (semver.Prerelease(v) == "" || includePrerelease) {
+			latest = v
+		}
+	}
+	return latest, nil
+}
+
+// GetModuleVersions retrieve a list of module versions for the specified module by querying the list
+// of module proxies configured on p.
+func (p Proxy) GetModuleVersions(mod string) ([]string, error) {
+	for _, proxy := range p.proxies {
+		url := proxy + "/" + path.Join(mod, "@v/list")
```

Comment:
Any escaping of `mod` we should care about here or does it always come in properly formatted?

Reply:
Probably fine to leave it without a check because any properly formatted module path will be valid here.